distribution/tools/server-cli/src/main/java/org/elasticsearch/server/cli/APMJvmOptions.java

```diff
@@ -0,0 +1,279 @@
+        STATIC_CONFIG.put("service_version", Version.CURRENT.toString());
+
+        // Configures a log file to write to. `_AGENT_HOME_` is a placeholder used
+        // by the agent. Don't disable writing to a log file, as the agent will then
+        // require extra Security Manager permissions when it tries to do something
+        // else, and it's just painful.
+        STATIC_CONFIG.put("log_file", "_AGENT_HOME_/../../logs/apm.log");
+
+        // ES does not use auto-instrumentation.
+        STATIC_CONFIG.put("instrument", "false");
+
+        CONFIG_DEFAULTS = new HashMap<>();
+
+        // This is used to keep all the errors and transactions of a service
+        // together and is the primary filter in the Elastic APM user interface.
+        //
+        // You can optionally also set `service_node_name`, which is used to
+        // distinguish between different nodes of a service, therefore it should
+        // be unique for each JVM within a service. If not set, data
+        // aggregations will be done based on a container ID (where valid) or on
+        // the reported hostname (automatically discovered or manually
+        // configured through hostname). However, if this node's `node.name` is
+        // set, then that value is used for the `service_node_name`.
+        CONFIG_DEFAULTS.put("service_name", "elasticsearch");
+
+        // An arbitrary string that identifies this deployment environment. For
+        // example, "dev", "staging" or "prod". Can be anything you like, but must
+        // have the same value across different systems in the same deployment
+        // environment.
+        CONFIG_DEFAULTS.put("environment", "dev");
+
+        // Logging configuration. Unless you need detailed logs about what the APM
+        // is doing, leave this value alone.
+        CONFIG_DEFAULTS.put("log_level", "error");
+        CONFIG_DEFAULTS.put("application_packages", "org.elasticsearch,org.apache.lucene");
+        CONFIG_DEFAULTS.put("metrics_interval", "120s");
+        CONFIG_DEFAULTS.put("breakdown_metrics", "false");
+        CONFIG_DEFAULTS.put("central_config", "false");
+    }
+
+    /**
+     * This method works out if APM tracing is enabled, and if so, prepares a temporary config file
+     * for the APM Java agent and CLI options to the JVM to configure APM. The config file is temporary
+     * because it will be deleted once Elasticsearch starts.
+     *
+     * @param settings the Elasticsearch settings to consider
+     * @param keystore a wrapper to access the keystore, or null if there is no keystore
+     * @param tmpdir Elasticsearch's temporary directory, where the config file will be written
+     */
+    static List<String> apmJvmOptions(Settings settings, @Nullable KeyStoreWrapper keystore, Path tmpdir) throws UserException,
+        IOException {
+        final boolean enabled = settings.getAsBoolean("xpack.apm.enabled", false);
```

Comment:
Is this a dynamic setting (cluster setting) or node local? I thought from reading other code it was dynamic, so it could be disabled on the fly.